include/beast/core/handler_ptr.hpp

```diff
@@ -0,0 +1,156 @@
+//
+// Copyright (c) 2013-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_HANDLER_PTR_HPP
+#define BEAST_HANDLER_PTR_HPP
+
+#include <beast/core/detail/type_traits.hpp>
+#include <atomic>
+#include <cstdint>
+#include <utility>
+
+namespace beast {
+
+namespace detail {
+
+template<class T, class = beast::detail::void_t<>>
+struct has_handler : std::false_type {};
+
+template<class T>
+struct has_handler<T, beast::detail::void_t<
+    decltype(T::handler)
+        > > : std::true_type {};
+
+} // detail
+
+/** A smart pointer container.
+
+    This is a smart pointer that retains shared ownership of an
+    object through a pointer. Memory is managed using the allocation
+    and deallocation functions associated with a completion handler,
+    which is also stored in the object. The object is destroyed and
+    its memory deallocated when one of the following happens:
+
+    * The last remaining container owning the object is destroyed
+    * The last remaining container owning the object is assigned
+      another object via @ref operator= or @ref reset.
```

Comment:
Remove this line? It looks like both `operator=` and `reset` functions were removed.